src/native/eventpipe/ep-session.c

```diff
@@ -63,9 +69,13 @@ EP_RT_DEFINE_THREAD_FUNC (streaming_thread)
 				ep_rt_wait_event_wait (wait_event, EP_INFINITE_WAIT, false);
 			}
 
-			// Wait until it's time to sample again.
-			const uint32_t timeout_ns = 100000000; // 100 msec.
-			ep_rt_thread_sleep (timeout_ns);
+			stop = ep_rt_perf_counter_query ();
```

Comment:
The intent is for us to stick to the original intent of "flush once every 100 ms". Before this patch, we were doing work and then sleeping 100ms regardless of how long we worked for.

This change moves the deletion outside the lock and adds a time budget. Now we do the same amount of work but stick to the original intent of "once per 100 ms" due to the budgeting.

Technically, yes, this will result in the thread working more often, but it should be doing the same amount of work. I think this was the original intent all along.

Reply:
Yes, but I'm not sure if the "intent" for this 100ms sleep is so that it flushes every 100ms regularly, or it's so that we let the session collect additional data for 100ms before we flush again so that we're not doing the expensive "flush all" too often esp. for low-verbosity sessions (i.e. counter sessions?).

Thinking about it from first principles, for low-verbosity sessions the write itself should be quite cheap and is likely the behavior here doesn't change that drastically, but I'd like to see some numbers just to make sure :p 

For verbose sessions I think this change is doing the right thing 👍

Reply:
For low verbosity sessions, this should be hitting the `!events_written` case and waiting on the WaitEvent until an event is written.

I can collect some numbers though to be sure we aren't introducing unnecessary CPU usage.

Reply:
The sleep is trying to strike a balance between two competing goals:
1. Low latency - if some event occurs at time T, we want that event to appear in the trace output within a short period of time so that the consumption experience appears immediate.
2. Efficient buffering - There is some work involved in allocating and deallocating new buffers, handoff between threads, and serializing them that is fixed cost regardless of how many events are in a buffer. So for best CPU and stream serialization size efficiency we want as many events in the same buffer as possible.

This is identical to what the [Nagle algorithm](https://en.wikipedia.org/wiki/Nagle's_algorithm#:~:text=Nagle%27s%20algorithm%20is%20a%20means%20of%20improving%20the,Control%20in%20IP%2FTCP%20Internetworks%20%28see%20RFC%20896%20%29.) is trying to solve for TCP communication.

@sywhang is right that (as written) this is going to sacrifice the efficient buffering goal for low verbosity channels. For example imagine an app that writes 10 events with a small delay between each(~1ms), then goes quiet for 60 seconds before repeating (EventCounter might look like this). ep_session_write_all_buffers_to_file() reports no events written, thread goes to sleep, wakes up ~60 second later when the first event is written. The timestamps will report worked_ms = 60,000, no sleeping will happen, loop wraps around and ep_session_write_all_buffers_to_file() writes out a single event. Now worked_ms = 0, sleep for 100ms occurs, the other 9 events queue up and they are written out in the 3rd iteration of the loop. What we wanted to see was all 10 events in a single file buffer but instead we get two buffers each time, one with a single event and one with 9 events.

I think you just need to ensure that worked_ms = 0 if this loop iteration blocked in ep_rt_wait_event_wait